### ISSUE TITLE
[FEATURE] V2 Volume Supports Cloning

### ISSUE THREAD

**Issue by draghuram** [longhorn/longhorn#7794]
Starting from Longhorn 1.3, when "type: snap" is configured in volume snapshot class, CSI snapshot requests (creation of VolumeSnapshot resource) results in a point in time Longhorn snapshot. 

However, if a PVC is created with CSI snapshot as data source, a new Longhorn volume is created and data is copied to it from the snapshot. Because of the copy, PVC is not ready immediately and when it becomes ready depends on the amount of data being copied. More over, the new PVC will by default have same number of replicas as the source volume but this can be fixed by using a different storage class that sets number of replicas to 1. While the data copy is fine in many cases, there is one particular use case where it is problematic.

Backup applications (such as CloudCasa and Velero) typically create a PVC snapshot and then backup its content. To read data from snapshot, they create a temporary PVC from the snapshot and read from it. Once backup is done, the temporary PVC is deleted. In this particular case, it is inefficient to have data copy take place when PVC is created from snapshot. Ideally, new volume should be created from snapshot with  no data copy as some storage systems do. If the new volume allows writes, modified  blocks should be saved separately while reads for other blocks can be fulfilled from the snapshot.

I don't know Longhorn internals  enough to tell whether this feature is feasible but I am opening this request after discussions with @PhanLe1010 in https://github.com/vmware-tanzu/velero/issues/6179#issuecomment-1882187472.

**Comment by innobead** [comment]
This could be a topic for v2 as well if we can share the immutable snapshot by reference in different volumes. cc @derekbit @shuo-wu @DamiaSan

**Comment by DamiaSan** [comment]
Instead of creating a new lvol and then copying snapshot's data into it, we could create a new clone of the snapshot.

**Comment by PhanLe1010** [comment]
TODO: 

Investigate if CDI flow + Longhorn CSI snapshot/fast cloning can have achieve similar goals as the Longhorn Backing Image flow

**Comment by PhanLe1010** [comment]
> Instead of creating a new lvol and then copying snapshot's data into it, we could create a new clone of the snapshot.

Thanks @DamiaSan  Looks like it is easier to be done in v2 data engine. Cool

**Comment by derekbit** [comment]
In favor of https://github.com/longhorn/longhorn/issues/9106

**Comment by ChanYiLin** [comment]
Hi @DamiaSan 
Suppose I have a volume with snapshot chain
“snap1->snap2->replica head”
When I clone from the volume, I will get
“snap1->snap2–>snap3-> [replca head, cloned lvol]”
Both clone lvol and replica head will have the same base_snapshot which is snap3

What will happen if I delete the volume, so the chain
“snap1->snap2->snap3->replica head” is deleted

Will the data of the snap1, snap2 and snap3 be merged to the cloned lvol?

**Comment by DamiaSan** [comment]
> Hi [@DamiaSan](https://github.com/DamiaSan) Suppose I have a volume with snapshot chain “snap1->snap2->replica head” When I clone from the volume, I will get “snap1->snap2–>snap3-> [replca head, cloned lvol]” Both clone lvol and replica head will have the same base_snapshot which is snap3
> 
> What will happen if I delete the volume, so the chain “snap1->snap2->snap3->replica head” is deleted
> 
> Will the data of the snap1, snap2 and snap3 be merged to the cloned lvol?

If you delete the volume, only replica head is deleted. Replica head, snap1,2,3 and cloned lvol are all logical volumes.

**Comment by ChanYiLin** [comment]
Hi @DamiaSan I think we not only delete the replica head but also the snapshot chain when deleting the volume

https://github.com/longhorn/longhorn-spdk-engine/blob/6e966aacbb0655b631e6ff481932493fef1c7603/pkg/spdk/replica.go#L952

**Comment by PhanLe1010** [comment]
Will cloned lvol be deleted if the original lvol is deleted? 
Maybe SPDK will block the deletion of the original lvol ?

**Comment by innobead** [comment]
> Sorry. I am still learning this but:
> 
> * Will cloned lvol be deleted if the original lvol is deleted?
> * Maybe SPDK will block the deletion of the original lvol ?

I would expect to, make the shared snapshot chain immutable after sharing with more than one volume replica instead. That might be another topic for space usage.

**Comment by DamiaSan** [comment]
> Hi [@DamiaSan](https://github.com/DamiaSan) I think we not only delete the replica head but also the snapshot chain when deleting the volume
> 
> https://github.com/longhorn/longhorn-spdk-engine/blob/6e966aacbb0655b631e6ff481932493fef1c7603/pkg/spdk/replica.go#L952

Sorry, I meant spdk side.
If the snapshot chain is deleted, then you are right, the data of snap1, snap2 and snap3 will be merged to the cloned lvol.

**Comment by DamiaSan** [comment]
> * Will cloned lvol be deleted if the original lvol is deleted?

SPDK side, not. Every snapshot, replica head and clone is an lvol, so every deletion involve only that lvol. Data will be merged to the child of the deleted lvol.

**Comment by PhanLe1010** [comment]
> SPDK side, not. Every snapshot, replica head and clone is an lvol, so every deletion involves only that lvol. Data will be merged to the child of the deleted lvol.

Thanks. 
What is the `child of the deleted lvol`? Do you mean the cloned lvol?

**Comment by DamiaSan** [comment]
> > SPDK side, not. Every snapshot, replica head and clone is an lvol, so every deletion involves only that lvol. Data will be merged to the child of the deleted lvol.
> 
> Thanks. What is the `child of the deleted lvol`? Do you mean the cloned lvol?

Yes, exactly.
If you have the following chain: `snap1 -> snap2 -> head lvol`
and you delete for example `snap2`, `snap2` data will be merged into `head lvol`.

**Comment by ChanYiLin** [comment]
Hi @PhanLe1010 
I am also thinking how to apply encryption when using the volume clone to support CDI path with encryption volume
In BackingImage, we do the encryption by creating another BackingImage.
1. Setup the loop device of the source BackingImageA and target BackingImageB
2. Add dm_encrypt layer to the target BackingImageB
3. Then copy the data from the source to the target to enrypt the data
4. The target BackingImageB will have encrypted data.

But if we are simply clone the snapshot lvol to get the volume clone (fast cloning)
We won't be able to do the encryption.

Maybe we will still need to rely on the traditional data copy way (CDI mount two volumes and do the copy) to support encrypted root volume?

cc @derekbit @innobead @bk201 @Vicente-Cheng @WebberHuang1118

**Comment by WebberHuang1118** [comment]
> Hi [@PhanLe1010](https://github.com/PhanLe1010) I am also thinking how to apply encryption when using the volume clone to support CDI path with encryption volume In BackingImage, we do the encryption by creating another BackingImage.
> 
> 1. Setup the loop device of the source BackingImageA and target BackingImageB
> 2. Add dm_encrypt layer to the target BackingImageB
> 3. Then copy the data from the source to the target to enrypt the data
> 4. The target BackingImageB will have encrypted data.
> 
> But if we are simply clone the snapshot lvol to get the volume clone (fast cloning) We won't be able to do the encryption.
> 
> Maybe we will still need to rely on the traditional data copy way (CDI mount two volumes and do the copy) to support encrypted root volume?
> 
> cc [@derekbit](https://github.com/derekbit) [@innobead](https://github.com/innobead) [@bk201](https://github.com/bk201) [@Vicente-Cheng](https://github.com/Vicente-Cheng) [@WebberHuang1118](https://github.com/WebberHuang1118)

Hi Jack,

If using the CDI way, IIUC, the flow will be like this:
- CDI creates an empty LH encryption volume (let's say `golden-encrypt-pvc`) with encryption sc
- CDI puts source image data into `golden-encrypt-pvc` (Harvester will wrappers `golden-encrypt-pvc` as a VM image to users)
- Once the user creates a new PVC from `golden-encrypt-pvc`
  - If the storage provider supports snapshot:
    - CDI will use take `VolumeSnapshot` on `golden-encrypt-pvc`, and create a new PVC from the snapshot
  - If the storage provider doesn't support snapshots:
    - CDI will use the copy way, i.e. copy data from `golden-encrypt-pvc` to the new PVC
   
My question is if CDI uses the `snapshot way`, does LH keep the data consistency as an encrypted volume?

@Vicente-Cheng please correct me, if my understanding is not right, thanks

**Comment by ChanYiLin** [comment]
> CDI puts source image data into golden-encrypt-pvc

If CDI can handle mounting the golden-encrypt-pvc with the dm_crypt, then when storing the data to that pvc, the data will be encrypted.
Following question:
Q: Is it possible for CDI to support creating a godlen-encrypte-pvc from an existing Volume?

 > My question is if CDI uses the snapshot way, does LH keep the data consistency as an encrypted volume?
Yes, I think taking the snapshot from an existing volume maintains the data consistency.

**Comment by WebberHuang1118** [comment]
> > CDI puts source image data into golden-encrypt-pvc
> 
> If CDI can handle mounting the golden-encrypt-pvc with the dm_crypt, then when storing the data to that pvc, the data will be encrypted. 

If `golden-encrypt-pvc` is created with encryption sc (like the sc in this [page](https://longhorn.io/docs/1.8.1/advanced-resources/security/volume-encryption/)), IIUC, the dm_crypt layer is already handled by LH, right? Why CDI has to handle the dm-crypt layer?

**Comment by ChanYiLin** [comment]
Ok got it. Then this path should work.

Then the question is 
Is it possible for CDI to support creating a godlen-encrypte-pvc from an existing Volume?

**Comment by ChanYiLin** [comment]
After discussed with @WebberHuang1118 , I am more clear now
1. No matter the source, when preparing the golden pvc in CDI, it launches the pod and mount the target PVC and do the data copy to store the data into that PVC. **We can do the encryption here by adding encryption to the target PVC**
    - The source can be a URL. CDI downloads the data to the target PVC.
    - The source can be another PVC. CDI mount the source PVC then dump the data to the target PVC
2. When creating a root disk from the golden PVC, this is where the smart clone gets involved. It creates a `volumeSnapshot` from the golden PVC and create another PVC from that `volumeSnapshot`
    - Underneath, Longhorn implements the smart clone by utilizing the spdk snapshot clone.

```
= smart clone = 
- use snapshot clone from spdk to prevent data copy

golden PVC
-> root diskA
-> root diskB
---
= Export= 
- original volume as the source, CDI launch a pod to mount two PVCs to do the full data copy

golden PVCA -> snapshot1 -> snapshot2 -> volumeA-head
golden PVCB
```

cc @WebberHuang1118 @Vicente-Cheng correct me if I misunderstand something, thanks!

**Comment by derekbit** [comment]
Let's focus on v2 data engine in the ticket. 
cc @PhanLe1010 @ChanYiLin @innobead

**Comment by PhanLe1010** [comment]
> Let's focus on v2 data engine in the ticket. cc [@PhanLe1010](https://github.com/PhanLe1010) [@ChanYiLin](https://github.com/ChanYiLin) [@innobead](https://github.com/innobead)

I agree. I changed the ticket subject to refer the v2 data engine specifically and use this ticket to continue forward. WDYT?

**Comment by derekbit** [comment]
> > Let's focus on v2 data engine in the ticket. cc [@PhanLe1010](https://github.com/PhanLe1010) [@ChanYiLin](https://github.com/ChanYiLin) [@innobead](https://github.com/innobead)
> 
> I agree. I changed the ticket subject to refer the v2 data engine specifically and use this ticket to continue forward. WDYT?

No prob!

**Comment by PhanLe1010** [comment]
Fyi, I think this feature can also serve as the foundation for v2 volume exporting/dowloading

**Comment by PhanLe1010** [comment]
Update: I will talk more about this in the LEP, but the current thought is that this ticket has 2 goals:
1. Regular cloning in which the data are decoupled from the source volume. The newly target volume can have multiple replicas. And the replica of the target volume can live in different node/disk as the source volume
1. Fast cloning in which data are depend on the source volume. The target volume cannot have multiple replica. And the replica of the target volume must be on same disk as one of the replica of the source volume. Main use case for this is external backup operation like Kasten or Velero which only need to read the data once and delete the cloned volume. There should be zero data copy in this case and clone should be instant

**Comment by innobead** [comment]
@derekbit as discussed, moving to 1.10.0, but we can backport it to 1.9.1 when available, as v2 is still in experimental stage. cc @PhanLe1010

**Comment by PhanLe1010** [comment]
**Update:**

We are going to support 2 cloning modes for v2 data engine:
* `full-copy` mode: efficiently copy the data allocated in the source volume to the target volume. Skipped the unallocated data. Keep the target volume thinly provisioned.
* `linked-clone` mode: No data moving. Instantly fast clone. Data still linked to the source volume snapshot. Purpose is for external backup solution such as velero or kasten which need to quickly clone and read data out of the snapshot then delete the target volume.

PRs are ready for review

**Comment by draghuram** [comment]
Thanks. Just wanted to mention that CloudCasa will be able to make use of this feature as well (that is the reason I opened the request in the first place).

**Comment by PhanLe1010** [comment]
## Test Plan:
(in-progress)

### Case 1: full-clone; src volume is attached; dst volume single replica
1. Create `src-vol` of 16 Mi
1. Attach `src-vol`. Write data to cluster 1 and 4 by `dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=1 count=1 conv=notrunc oflag=direct && dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=4 count=1 conv=notrunc oflag=direct`
1. Take a new snapshot
1. Write data to cluster 4 and 7 by `dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=4 count=1 conv=notrunc oflag=direct && dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=7 count=1 conv=notrunc oflag=direct`
1. Using UI or [yaml](https://longhorn.io/docs/1.9.1/snapshots-and-backups/csi-volume-clone/) to clone the volume call `dst-vol`. `dst-vol` should have 1 replica
1. Wait for cloning to finish
1. Attach the `dst-vol`. 
1. Make sure they have same checksum 
1. Make sure the actual size of `dst-vol` is 3 Mi


### Case 2: full-clone; src volume is detached; dst volume single replica
1. Create `src-vol` of 16 Mi
1. Attach `src-vol`. Write data to cluster 1 and 4 by `dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=1 count=1 conv=notrunc oflag=direct && dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=4 count=1 conv=notrunc oflag=direct`
1. Take a new snapshot
1. Write data to cluster 4 and 7 by `dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=4 count=1 conv=notrunc oflag=direct && dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=7 count=1 conv=notrunc oflag=direct`
1. Detach the `src-vol`
1. Using UI or [yaml](https://longhorn.io/docs/1.9.1/snapshots-and-backups/csi-volume-clone/) to clone the volume call `dst-vol`. `dst-vol` should have 1 replica
1. Wait for cloning to finish
1. Attach the `dst-vol`. 
1. Make sure they have same checksum 
1. Make sure the actual size of `dst-vol` is 3 Mi

### Case 3: full-clone; src volume is detached; dst volume multiple replicas
1. Create `src-vol` of 16 Mi
1. Attach `src-vol`. Write data to cluster 1 and 4 by `dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=1 count=1 conv=notrunc oflag=direct && dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=4 count=1 conv=notrunc oflag=direct`
1. Take a new snapshot
1. Write data to cluster 4 and 7 by `dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=4 count=1 conv=notrunc oflag=direct && dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=7 count=1 conv=notrunc oflag=direct`
1. Detach the `src-vol`
1. Using UI or [yaml](https://longhorn.io/docs/1.9.1/snapshots-and-backups/csi-volume-clone/) to clone the volume call `dst-vol`. `dst-vol` should have 3 replicas
1. Wait for cloning to finish
1. Verify that `dst-vol` has 3 healthy replicas before the clone is finished 
1. Attach the `dst-vol`. 
1. Make sure they have same checksum 
1. Make sure the actual size of `dst-vol` is 3 Mi


### Case 4: linked-clone; src volume is attached; dst volume single replica
1. Create `src-vol` of 16 Mi
1. Attach `src-vol`. Write data to cluster 1 and 4 by `dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=1 count=1 conv=notrunc oflag=direct && dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=4 count=1 conv=notrunc oflag=direct`
1. Using UI or [yaml](https://longhorn.io/docs/1.9.1/snapshots-and-backups/csi-volume-clone/) to clone the volume  `src-vol` to create `dst-vol` with 1 replica and `linked-clone` mode
1. Wait for cloning to finish
1. Attach the `dst-vol` and verify the data checksum

### Case 5: linked-clone; src volume is detached; dst volume single replica
Same as case 4 but detach the `src-vol` before doing cloning

### Case 6: linked-clone; src volume has big actual size
Same as case 4 but use `src-vol` with 5GiB of actual size. Verify that the clone finished within 10s

### Case 7: linked-clone; src volume is detached; dst volume multiple replicas
1. Create `src-vol` of 16 Mi
1. Attach `src-vol`. Write data to cluster 1 and 4 by `dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=1 count=1 conv=notrunc oflag=direct && dd if=/dev/urandom of=/dev/longhorn/src-vol bs=1M seek=4 count=1 conv=notrunc oflag=direct`
1. Using UI or [yaml](https://longhorn.io/docs/1.9.1/snapshots-and-backups/csi-volume-clone/) to clone the volume  `src-vol` to create `dst-vol` with 2 replica and `linked-clone` mode
1. See that the creation failed because linked-clone `dst-vol` can only has 1 replica

### Case 8: linked-clone; src volume already has a linked-clone volume
1. Create `src-vol` of 16 Mi
1. Using UI or [yaml](https://longhorn.io/docs/1.9.1/snapshots-and-backups/csi-volume-clone/) to clone the volume  `src-vol` to create `dst-vol` with 1 replica and `linked-clone` mode
1. Wait for cloning to finish
1. Using UI or [yaml](https://longhorn.io/docs/1.9.1/snapshots-and-backups/csi-volume-clone/) to clone the volume  `src-vol` to create `dst-vol-2` with 1 replica and `linked-clone` mode
1. Verify that Longhorn reject the `dst-vol-2` creation. (There should be only 1 linked-clone volume for a src volume at a time. User has to delete `dst-vol` before being able to create `dst-vol-2`)

### Case 9: linked-clone; src volume is a linked-clone volume
1. Create `src-vol` of 16 Mi
1. Using UI or [yaml](https://longhorn.io/docs/1.9.1/snapshots-and-backups/csi-volume-clone/) to clone the volume `src-vol` to create `dst-vol` with 1 replica and `linked-clone` mode
1. Wait for cloning to finish
1. Using UI or [yaml](https://longhorn.io/docs/1.9.1/snapshots-and-backups/csi-volume-clone/) to clone the volume`dst-vol` to create `dst-vol-2` with 1 replica and `linked-clone` mode
1. Verify that Longhorn reject the `dst-vol-2` creation. (Cannot clone a volume from a linked-clone volume)

### Case 10: linked-clone; negative tests
Verify that `dst-vol` with `linked-clone` mode:
1. Cannot have more than 1 replica
1. Cannot take backup, snapshot
2. Cannot rebuild
3. Cannot change clone mode
4. Verify cannot add recurring job to linked-clone volume
    1. Create a default recurring job
    2. Create a linked-clone volume
    3. Try to add recurring job, it should fail
    4. Verify that default recurring job does not pickup the linked-clone volume when it run

**Comment by yangchiu** [comment]
Verification passed on master-head (longhorn-manager 2025/09/11 14:00 GMT+8) and v1.10.x-head (longhorn-manager 09c58af) following [the test plan](https://github.com/longhorn/longhorn/issues/7794#issuecomment-3221870866).

For `Case 1` ~ `Case 3`, use the following custom storage class to full copy the PVC:

```
kubectl apply -f - <<EOF
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: full-copy-sc
provisioner: driver.longhorn.io
reclaimPolicy: Delete
volumeBindingMode: Immediate
parameters:
  numberOfReplicas: "1"
  staleReplicaTimeout: "2880"
  fsType: "ext4"
  dataEngine: "v2"
  cloneMode: "full-copy"
EOF
```

```
kubectl apply -f - <<EOF
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: dst-vol
spec:
  storageClassName: full-copy-sc
  dataSource:
    name: src-vol
    kind: PersistentVolumeClaim
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 16Mi
EOF
```

For `Case 4` ~ `Case 10`, use the following custom storage class to linked clone the PVC:

```
kubectl apply -f - <<EOF
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: linked-clone-sc
provisioner: driver.longhorn.io
reclaimPolicy: Delete
volumeBindingMode: Immediate
parameters:
  numberOfReplicas: "1"
  staleReplicaTimeout: "2880"
  fsType: "ext4"
  dataEngine: "v2"
  cloneMode: "linked-clone"
EOF
```

```
kubectl apply -f - <<EOF
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: dst-vol
spec:
  storageClassName: linked-clone-sc
  dataSource:
    name: src-vol
    kind: PersistentVolumeClaim
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 16Mi
EOF
```

For `Case 7: linked-clone; src volume is detached; dst volume multiple replicas`, the cloning failed with the following error message:
```
$ kubectl describe pvc dst-vol
Name:          dst-vol
Namespace:     default
StorageClass:  linked-clone-sc
Status:        Pending
Volume:        
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: driver.longhorn.io
               volume.kubernetes.io/storage-provisioner: driver.longhorn.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
DataSource:
  Kind:   PersistentVolumeClaim
  Name:   src-vol
Used By:  <none>
Events:
  Type     Reason                Age               From                                                                                      Message
  ----     ------                ----              ----                                                                                      -------
  Normal   ExternalProvisioning  9s (x3 over 16s)  persistentvolume-controller                                                               Waiting for a volume to be created either by the external provisioner 'driver.longhorn.io' or manually by the system administrator. If volume creation is delayed, please verify that the provisioner is running and correctly registered.
  Warning  ProvisioningFailed    9s (x4 over 16s)  driver.longhorn.io_csi-provisioner-7cd5b454df-5bh4f_0cff99d5-c631-4fb9-acb2-50f667cee1a6  failed to provision volume with StorageClass "linked-clone-sc": rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [code=Internal Server Error, detail=, message=failed to create volume: unable to create volume pvc-561d5b99-a267-4c44-be26-9a896784d827: admission webhook "validator.longhorn.io" denied the request: number of replica count must be 1 when clone mode linked-clone] from [http://longhorn-backend:9500/v1/volumes]
  Normal   Provisioning          1s (x5 over 16s)  driver.longhorn.io_csi-provisioner-7cd5b454df-5bh4f_0cff99d5-c631-4fb9-acb2-50f667cee1a6  External provisioner is provisioning volume for claim "default/dst-vol"
  Warning  ProvisioningFailed    1s                driver.longhorn.io_csi-provisioner-7cd5b454df-5bh4f_0cff99d5-c631-4fb9-acb2-50f667cee1a6  failed to provision volume with StorageClass "linked-clone-sc": rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [message=failed to create volume: unable to create volume pvc-561d5b99-a267-4c44-be26-9a896784d827: admission webhook "validator.longhorn.io" denied the request: number of replica count must be 1 when clone mode linked-clone, code=Internal Server Error, detail=] from [http://longhorn-backend:9500/v1/volumes]
```

For `Case 8 linked-clone; src volume already has a linked-clone volume`, the cloning failed with the following error message:
```
$ kubectl describe pvc dst-vol-2
Name:          dst-vol-2
Namespace:     default
StorageClass:  linked-clone-sc
Status:        Pending
Volume:        
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: driver.longhorn.io
               volume.kubernetes.io/storage-provisioner: driver.longhorn.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
DataSource:
  Kind:   PersistentVolumeClaim
  Name:   src-vol
Used By:  <none>
Events:
  Type     Reason                Age               From                                                                                      Message
  ----     ------                ----              ----                                                                                      -------
  Warning  ProvisioningFailed    14s               driver.longhorn.io_csi-provisioner-7cd5b454df-5bh4f_0cff99d5-c631-4fb9-acb2-50f667cee1a6  failed to provision volume with StorageClass "linked-clone-sc": rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [message=failed to create volume: unable to create volume pvc-ee9f0c56-461b-45e1-8304-8d2906dd6b58: admission webhook "validator.longhorn.io" denied the request: BUG: there already exist a linked-cloned volume pvc-086f7fc4-246e-4ba1-bcce-7595a4f8e1c4 from the source volume src-vol, code=Internal Server Error, detail=] from [http://longhorn-backend:9500/v1/volumes]
  Normal   Provisioning          8s (x4 over 15s)  driver.longhorn.io_csi-provisioner-7cd5b454df-5bh4f_0cff99d5-c631-4fb9-acb2-50f667cee1a6  External provisioner is provisioning volume for claim "default/dst-vol-2"
  Warning  ProvisioningFailed    8s (x3 over 15s)  driver.longhorn.io_csi-provisioner-7cd5b454df-5bh4f_0cff99d5-c631-4fb9-acb2-50f667cee1a6  failed to provision volume with StorageClass "linked-clone-sc": rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [code=Internal Server Error, detail=, message=failed to create volume: unable to create volume pvc-ee9f0c56-461b-45e1-8304-8d2906dd6b58: admission webhook "validator.longhorn.io" denied the request: BUG: there already exist a linked-cloned volume pvc-086f7fc4-246e-4ba1-bcce-7595a4f8e1c4 from the source volume src-vol] from [http://longhorn-backend:9500/v1/volumes]
  Normal   ExternalProvisioning  2s (x3 over 15s)  persistentvolume-controller                                                               Waiting for a volume to be created either by the external provisioner 'driver.longhorn.io' or manually by the system administrator. If volume creation is delayed, please verify that the provisioner is running and correctly registered.
```

For `Case 9: linked-clone; src volume is a linked-clone volume`, the cloning failed with the following error message:
```
$ kubectl describe pvc dst-vol-3
Name:          dst-vol-3
Namespace:     default
StorageClass:  linked-clone-sc
Status:        Pending
Volume:        
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: driver.longhorn.io
               volume.kubernetes.io/storage-provisioner: driver.longhorn.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
DataSource:
  Kind:   PersistentVolumeClaim
  Name:   dst-vol-2
Used By:  <none>
Events:
  Type     Reason                Age               From                                                                                      Message
  ----     ------                ----              ----                                                                                      -------
  Normal   Provisioning          6s (x4 over 13s)  driver.longhorn.io_csi-provisioner-7cd5b454df-5bh4f_0cff99d5-c631-4fb9-acb2-50f667cee1a6  External provisioner is provisioning volume for claim "default/dst-vol-3"
  Warning  ProvisioningFailed    6s (x4 over 13s)  driver.longhorn.io_csi-provisioner-7cd5b454df-5bh4f_0cff99d5-c631-4fb9-acb2-50f667cee1a6  failed to provision volume with StorageClass "linked-clone-sc": rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [code=Internal Server Error, detail=, message=failed to create volume: unable to create volume pvc-414b8ed3-005f-44ba-8de8-f7232fc43a05: admission webhook "validator.longhorn.io" denied the request: cannot create a new volume from a linked-clone volume pvc-ee9f0c56-461b-45e1-8304-8d2906dd6b58] from [http://longhorn-backend:9500/v1/volumes]
  Normal   ExternalProvisioning  5s (x3 over 13s)  persistentvolume-controller                                                               Waiting for a volume to be created either by the external provisioner 'driver.longhorn.io' or manually by the system administrator. If volume creation is delayed, please verify that the provisioner is running and correctly registered.
```

**Comment by PhanLe1010** [comment]
Ref the blog [Backup Applications with Longhorn V2 Volumes using Velero](https://longhorn.io/blog/20250902-k8s-backup-solutions-and-longhorn/) about how to integrate this feature. cc @draghuram Maybe CloudCasa will behave the similarly?

**Comment by draghuram** [comment]
Yes, CloudCasa will follow similar steps for backup where it creates a temporary PVC from snapshot. So if the storage class is properly configured, the operation should be quick. We will try to verify it.

**Comment by c3y1huang** [comment]
@PhanLe1010 I am re-opening this issue as it’s missing documentation and important note.

cc @derekbit @longhorn/qa

**Comment by PhanLe1010** [comment]
Doc PR pushed https://github.com/longhorn/website/pull/1201

**Comment by derekbit** [comment]
Completed.

**Comment by chriscchien** [comment]
Verified **pass** on longhorn v1.11.x-head(https://github.com/longhorn/longhorn-manager/commit/e9aeeb9b3824f745eb092dac2cf69a6a1a9dc611) with [test steps](https://github.com/longhorn/longhorn/issues/7794#issuecomment-3221870866)
